src/crimson/osd/pg.cc

```diff
@@ -903,23 +921,26 @@ PG::do_osd_ops_execute(
       std::move(all_completed_fut)
     );
   }, OpsExecuter::osd_op_errorator::all_same_way(
-    [this, rollbacker, failure_func_ptr]
+    [this, error_func_ptr, rollbacker, failure_func_ptr]
     (const std::error_code& e) mutable {
 
-    auto submitted_fut = seastar::now();
+    PG::interruptible_future<> maybe_rollback_fut = seastar::now();
     ceph_tid_t rep_tid = shard_services.get_tid();
 
-    auto all_completed_fut = e.value() == ENOENT ?
-      (*failure_func_ptr)(e, rep_tid) :
-      rollbacker.rollback_obc_if_modified(e).then_interruptible(
-      [e, failure_func_ptr, rep_tid] {
-          return (*failure_func_ptr)(e, rep_tid);
-      });
+    if (e.value() == ENOENT) {
+      maybe_rollback_fut = rollbacker.rollback_obc_if_modified(e);
+    }
 
-    return PG::do_osd_ops_iertr::make_ready_future<pg_rep_op_fut_t<Ret>>(
-      std::move(submitted_fut),
-      std::move(all_completed_fut)
-    );
+    return maybe_rollback_fut.then_interruptible(
+    [error_func_ptr, e, rep_tid, failure_func_ptr] {
+      return (*error_func_ptr)(e, rep_tid).then(
+      [failure_func_ptr, e, rep_tid] {
+        return PG::do_osd_ops_iertr::make_ready_future<pg_rep_op_fut_t<Ret>>(
+          std::move(seastar::now()),
+          std::move((*failure_func_ptr)(e, rep_tid))
```

Comment:
1/2: This part lets `submit_error_log` run from within `do_osd_ops_execute` which resolves the original issue.

Reply:
This one is chained after the future returned from submit_transaction, which would otherwise resolve into the submitted and completed futures.  This one, I think, as you observe is the one that resolves the bug.

Reply:
It seems to me that error_func_ptr could return an eversion_t, pass it into failure_func_ptr, and thereby avoid the log_entry_version map.

If we retain the log_entry_version map, it probably needs to be cleared during interval change.

Reply:
> It seems to me that error_func_ptr could return an eversion_t, pass it into failure_func_ptr, and thereby avoid the log_entry_version map.

After the last commit, it is no longer possible.

> If we retain the log_entry_version map, it probably needs to be cleared during interval change.
 
Fixed.